### PR TITLE
Add GNN Example

### PR DESCRIPTION
**Pull request by wangxiaoyunNV**
Fixes # .

### Description
Added GNN example with [graphsage](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GraphSAGE.html) model, using PyG as a backend for GNN. In this example, I use Protein-Protein Interaction dataset. I follow the implementation in PyG [example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup_ppi.py) for the GraphSage model. 
A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Comment by holgerroth** [comment]
Thanks for the contribution. Should this be in the "advanced" folder rather than "hello-world"?

File: examples/hello-world/gnn-pt/README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# Hello PyTorch
```
[the remark this answers is not in the document]
Comment (holgerroth) [reply examples/hello-world/gnn-pt/README.md:1]: Also, the content needs to be updated. Is there any data preprocessing required?

File: examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+from nvflare.apis.executor import Executor
+from nvflare.apis.fl_constant import ReservedKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.model import make_model_learnable, model_learnable_to_dxo
+from nvflare.app_common.app_constant import AppConstants
+from nvflare.app_opt.pt.model_persistence_format_manager import PTModelPersistenceFormatManager
+import torch.nn.functional as F
+from torch_geometric.loader import DataLoader,LinkNeighborLoader
```
Comment (holgerroth) [inline examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py:32]: `torch_geometric` and other dependencies need to be added to the requirements.txt.

File: examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/pt_model_locator.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+class PTModelLocator(ModelLocator):
+    def __init__(self):
+        super().__init__()
+
+        path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
+        train_dataset = PPI(path, split='train')
+        train_data = Batch.from_data_list(train_dataset)
+        self.model = GraphSAGE(
+            in_channels=train_dataset.num_features,
```
Comment (YuanTingHsieh) [inline examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/pt_model_locator.py:39]: These lines are used to determine the "in_channels" of the "GraphSAGE" model.

However, you are assuming the dataset also exists on server side.

There should be a way to send this information to server or pre-configured it.
@holgerroth @ZiyueXu77 can you advise on how you do this, thanks

File: examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+from torch_geometric.datasets import PPI
+from torch_geometric.data import Batch
+from torch_geometric.nn import GraphSAGE
+import tqdm
```
Comment (YuanTingHsieh) [inline examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py:36]: make sure `tqdm` is in "requirements.txt" as well

**Comment by YuanTingHsieh** [comment]
> Thanks for the contribution. Should this be in the "advanced" folder rather than "hello-world"?

@wangxiaoyunNV ^ can you move your example from "hello-world" to "advanced"?

Thanks!

File: examples/hello-world/gnn-pt/requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+nvflare[PT]>=2.3.0
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply examples/hello-world/gnn-pt/requirements.txt:1]: nvflare[PT] has torch.

others are valid, also might need "tqdm"

[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/requirements.txt:1]: also need torch-sparse for torch_geometric 
I remember scikit-learn is included in the pytorch, I did not installed in my environment

[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply examples/hello-world/gnn-pt/requirements.txt:1]: sounds good, I think scikit-learn is not included in Pytorch, I just tried pip install torch and import sklearn after that does not work.

File: examples/hello-world/gnn-pt/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/README.md]: data preprocessing is automatically taken care of by the PyG functions, so we do not need to by ourselves.

File: examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py (hunk 4 above)
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py]: added

File: examples/hello-world/gnn-pt/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+You can find the running logs and results inside the simulator's workspace/simulate_job
+
+```bash
+$ ls /tmp/nvflare/simulate_job/
+app_server  app_site-1  app_site-2  log.txt
+
+```
+
+### 4. Global Model Initialization Approaches
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/README.md]: removed

File: examples/hello-world/gnn-pt/requirements.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/requirements.txt]: all added

File: examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py (hunk 2 above)
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/jobs/gnn-pt/app/custom/graphsagetrainer.py]: added

File: examples/hello-world/gnn-pt/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/hello-world/gnn-pt/README.md]: changed to graphsage

File: examples/advanced/gnn-pt/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+### 3. Access the logs and results
+
+You can find the running logs and results inside the simulator's workspace/simulate_job
+
+```bash
+$ ls /tmp/nvflare/simulate_job/
+app_server  app_site-1  app_site-2  log.txt
+
+```
+
```
Comment (holgerroth) [inline examples/advanced/gnn-pt/README.md:38]: Can you add some expected results and training curves here like in other examples, e.g. this [one](https://github.com/NVIDIA/NVFlare/blob/dev/examples/advanced/cifar10/cifar10-sim/README.md#32-centralized-training)?

File: examples/advanced/gnn-pt/jobs/gnn-pt/app/config/config_fed_server.json [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        "expected_data_kind": "WEIGHTS",
+        "aggregation_weights": {
+          "site-1": 1.0,
+          "site-2": 1.0
+        }
+      }
+    },
+    {
+      "id": "model_locator",
+      "path": "pt_model_locator.PTModelLocator",
```
Comment (holgerroth) [inline examples/advanced/gnn-pt/jobs/gnn-pt/app/config/config_fed_server.json:33]: @YuanTingHsieh isn't is possible to use the standard `PTFileModelLocator` from app_common here?

File: examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/pt_model_locator.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+        path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
+        train_dataset = PPI(path, split='train')
+        train_data = Batch.from_data_list(train_dataset)
+        self.model = GraphSAGE(
+            in_channels=train_dataset.num_features,
+            hidden_channels=64,
+            num_layers=2,
+            out_channels=64,
+        )
```
Comment (YuanTingHsieh) [inline examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/pt_model_locator.py:43]: @holgerroth custom logic here^

File: examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        self._pre_train_task_name = pre_train_task_name
+        self._submit_model_task_name = submit_model_task_name
+        self._exclude_vars = exclude_vars
+
+        # Create PPI dataset for training.
+        path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
+        train_dataset = PPI(path, split='train')
+        val_dataset = PPI(path, split='val')
+        test_dataset = PPI(path, split='test')
+        kwargs = {'batch_size': 1024, 'num_workers': 6, 'persistent_workers': True}
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py:76]: train is used in the dataloader, I removed the rest

File: examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        train_data = Batch.from_data_list(train_dataset)
+        self._train_loader = LinkNeighborLoader(train_data, batch_size=2048, shuffle=True,
+                            neg_sampling_ratio=1.0, num_neighbors=[10, 10],
+                            num_workers=6, persistent_workers=True)
+        #self._train_loader = DataLoader(train_dataset, batch_size=2)
+        self._n_iterations = len(self._train_loader)
+        # Evaluation loaders (one datapoint corresponds to a graph)
+        train_loader = DataLoader(train_dataset, batch_size=2)
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py:84]: I have removed the unnecessary code.

File: examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+from nvflare.app_opt.pt.model_persistence_format_manager import PTModelPersistenceFormatManager
+import torch.nn.functional as F
+from torch_geometric.loader import DataLoader,LinkNeighborLoader
+from torch_geometric.datasets import PPI
+from torch_geometric.data import Batch
+from torch_geometric.nn import GraphSAGE
+import tqdm
+
+
+class GraphSageTrainer(Executor):
```
[the remark this answers is not in the document]
Comment (wangxiaoyunNV) [reply examples/advanced/gnn-pt/jobs/gnn-pt/app/custom/graphsage_trainer.py:39]: I followed the hello world pt as an example. So I used the Executor, could you please point me to the example that uses Learner Pattern?

**Comment by wangxiaoyunNV** [comment]
discard old version

**Pull request closed without merging**